### PR TITLE
Update `jacobian_tape.jacobian` to correctly support non-NumPy tensors

### PR DESCRIPTION
**Pull request by thisac**
**Context:**
Tests are failing in #1360 due to a torch tensor being incorrectly cast to a NumPy array with dtype `np.object_` instead of `float`. This is due to the additions in #1291, where before the `tape.jacobian_tape.jacobian` method always created a NumPy array with dtype `float`. This change caused post-processed results being e.g. `torch.Tensor` objects to set the dtype of the jacobian results to `np.object_`, even if the actual results where floats.

**Description of the Change:**
The post-processed results `g` are cast to a NumPy array and, if the elements are automatically cast to floats in the process, `g` is set to this array instead.

```python
g_arr = np.array(g)
if g_arr.dtype == float:
    g = g_arr
```

Since a NumPy array is returned anyway, this change only changes the dtype if the results would be cast to floats when input into the `jac` array anyway.

**Benefits:**
Things work better with Torch tensors.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: pennylane/tape/qubit_param_shift.py [hunk 1]
```diff
@@ -175,7 +176,7 @@ def processing_fn(results):
                 array[float]: 1-dimensional array of length determined by the tape output
                 measurement statistics
             """
-            results = np.squeeze(results)
+            results = np.squeeze([toarray(res) for res in results])
```
Comment (josh146) [inline pennylane/tape/qubit_param_shift.py:179]: 🙌

Comment (josh146) [reply pennylane/tape/qubit_param_shift.py:179]: I wonder if it is worth putting a try-except around this?

Comment (josh146) [reply pennylane/tape/qubit_param_shift.py:179]: (@cvjjm - the tests you added to ensure object results are returned still passes, but wondering if this line will have any implications wrt your current workflow).

Comment (cvjjm) [reply pennylane/tape/qubit_param_shift.py:179]: Thanks for thinking of me!  I don't fully understand what made this change necessary, but from a quick test I could not see any negative impact.

**Review by josh146 — APPROVED** [review]

**Comment by antalszava** [comment]
With #1360 coming up in the pipeline, the plan is to add a test using the new `default.qubit.torch` device. We've experienced the issue there for `diff_method='parameter-shift'`.

**Comment by josh146** [comment]
Just curious what the plan is for this PR?

**Comment by thisac** [comment]
@antalszava and I are working on it. We'll add tests, and plan on getting this into the next release instead.

**Comment by antalszava** [comment]
Once #1360 is merged, undoing [the original commit](https://github.com/PennyLaneAI/pennylane/pull/1360#discussion_r685598180) that removed the testing for `"parameter-shift"` will provide a good integration test. It tests the pipeline for `default.qubit.torch`, the only core device we've experienced issues with.

File: pennylane/tape/jacobian_tape.py [hunk 2]
```diff
@@ -360,7 +360,7 @@ def device_pd(self, device, params=None, **options):
 
         # TODO: modify devices that have device Jacobian methods to
         # accept the quantum tape as an argument
-        jac = jacobian_method(self, **options.get("device_pd_options", dict()))
+        jac = jacobian_method(self, **options.get("device_pd_options", {}))
```
Comment (antalszava) [inline pennylane/tape/jacobian_tape.py:363]: Changed this because Codefactor was unhappy.

File: tests/devices/test_default_qubit_torch.py [hunk 3]
```diff
@@ -1334,7 +1334,7 @@ def circuit(a, b):
         assert torch.allclose(b.grad, 0.5 * torch.sin(b) * (1 - torch.cos(a)))
 
     @pytest.mark.parametrize("operation", [qml.U3, qml.U3.decomposition])
-    @pytest.mark.parametrize("diff_method", ["backprop", "finite-diff"])
+    @pytest.mark.parametrize("diff_method", ["backprop", "parameter-shift", "finite-diff"])
```
Comment (antalszava) [inline tests/devices/test_default_qubit_torch.py:1337]: This is a test case that tests the entire pipeline (integration test).

File: tests/tape/test_qubit_param_shift.py [hunk 4]
```diff
@@ -300,6 +300,33 @@ def test_variance_gradients_agree_finite_differences(self, mocker, tol):
         assert np.allclose(grad_A, grad_F1, atol=tol, rtol=0)
         assert np.allclose(grad_A, grad_F2, atol=tol, rtol=0)
 
+    def test_processing_function_torch(self, mocker, tol):
```
Comment (antalszava) [inline tests/tape/test_qubit_param_shift.py:303]: This is a test case that checks that the output of the processing function is a numpy array. It fails when the change in `pennylane/tape/jacobian_tape.py` is not issued.

**Review by antalszava — APPROVED** [review]

**Pull request merged**